### PR TITLE
refactor: iam expire func

### PR DESCRIPTION
**Pull request by sqin2019**
part of #104

### CODE REVIEW

File: pkg/handler/iam_handler.go [hunk 1]
```diff
@@ -106,11 +109,29 @@ func NewIAMHandler(ctx context.Context, organizationsClient, foldersClient, proj
 	return h, nil
 }
 
+// Cleanup removes expired IAM bindings added by AOD from the IAM policies of the resources in the request.
+func (h *IAMHandler) Cleanup(ctx context.Context, r *v1alpha1.IAMRequest) (nps []*v1alpha1.IAMResponse, retErr error) {
```
Comment (sqin2019) [inline pkg/handler/iam_handler.go:113]: Though we have talked about this, but I wonder what will be the best approach here @yolocs :
1. when cleanup triggered on a request, cleanups all expired the AOD binding in the policy. 
2. only cleanups bindings on the request that are expired.
3. only cleanups bindings on the request no matter if it is expired.

Comment (yolocs) [reply pkg/handler/iam_handler.go:113]: I think it should be a combination of 1 and 3. It's safe to clean up all expired bindings. For 3, worst case scenario, a user needs to reopen the PR or open a new one.

File: pkg/handler/iam_handler.go [hunk 2]
```diff
@@ -106,11 +109,29 @@ func NewIAMHandler(ctx context.Context, organizationsClient, foldersClient, proj
...
+	for _, p := range r.ResourcePolicies {
+		// time.Now() is a dummy parameter used to match the function signature.
+		np, err := h.handlePolicy(ctx, p, time.Now(), h.cleanupExpiredBindings)
```
Comment (yolocs) [inline pkg/handler/iam_handler.go:116]: I don't think we should ignore cleanup error if it's an explicit cleanup call.

File: pkg/handler/iam_handler.go [hunk 3]
```diff
@@ -181,37 +202,33 @@ func (h *IAMHandler) handlePolicy(ctx context.Context, p *v1alpha1.ResourcePolic
...
 
-// Remove expired bindings and add or update new bindings with expiration condition.
-func (h *IAMHandler) updatePolicy(ctx context.Context, p *iampb.Policy, bs []*v1alpha1.Binding, expiry time.Time) {
-	logger := logging.FromContext(ctx)
+// addAndCleanupPolicy adds new bindings with expiration condition, it also does
+// best effort cleanup which removes any expired AOD bindings, however any
+// errors encounterred during removal will be ignored and policy update for the
+// request will continue. Removal errors should be handled separately such as
+// in a global IAM cleanup.
+func (h *IAMHandler) addAndCleanupPolicy(ctx context.Context, p *iampb.Policy, bs []*v1alpha1.Binding, expiry time.Time) {
```
Comment (yolocs) [inline pkg/handler/iam_handler.go:210]: For simplicity, I'd just call this `addPolicyBindings`

File: pkg/handler/iam_handler.go [hunk 4]
```diff
@@ -181,37 +202,33 @@ func (h *IAMHandler) handlePolicy(ctx context.Context, p *v1alpha1.ResourcePolic
...
+	// Cleanup policy.
+	h.cleanupExpiredBindings(ctx, p, bs, expiry)
```
Comment (yolocs) [inline pkg/handler/iam_handler.go:212]: This func should return an error and log warn here instead.

File: pkg/handler/iam_handler.go [hunk 5]
```diff
@@ -247,6 +264,38 @@ func (h *IAMHandler) updatePolicy(ctx context.Context, p *iampb.Policy, bs []*v1
 	p.Version = 3
 }
 
+// cleanupExpiredBindings does best effort cleanup which removes any expired
+// AOD bindings, however any errors encounterred during removal will be ignored
+// Removal errors should be handled separately such as in a global IAM cleanup.
+// bs and expiry parameters are not used, they are here to match the
+// updatePolicy func signature.
+func (h *IAMHandler) cleanupExpiredBindings(ctx context.Context, p *iampb.Policy, bs []*v1alpha1.Binding, expiry time.Time) {
```
Comment (yolocs) [inline pkg/handler/iam_handler.go:272]:
```suggestion
func (h *IAMHandler) cleanupExpiredBindings(ctx context.Context, p *iampb.Policy, _ []*v1alpha1.Binding, _ time.Time) {
```

**Review by yolocs — CHANGES_REQUESTED** [review]

File: pkg/handler/iam_handler.go (hunk 4 above)
Comment (sqin2019) [reply pkg/handler/iam_handler.go]: Updated.

File: pkg/handler/iam_handler.go (hunk 2 above)
Comment (sqin2019) [reply pkg/handler/iam_handler.go]: Updated

File: pkg/handler/iam_handler.go (hunk 1 above)
Comment (sqin2019) [reply pkg/handler/iam_handler.go:113]: Updated.

File: pkg/handler/iam_handler.go [hunk 6]
```diff
@@ -245,6 +239,59 @@ func (h *IAMHandler) updatePolicy(ctx context.Context, p *iampb.Policy, bs []*v1
...
+		}
+
+		// Check the expiration using expression directly instead of the given
+		// expiry as they may be different.
+		expired, err := expired(b.Condition.Expression)
+		if err != nil {
+			// Continue policy update when there is error checking the AOD expiry and
+			// return the errors.
+			// Cleaning up expired AOD bindings here is best effort.
+			// We rely on a separate process to clean up AOD bindings.
```
Comment (yolocs) [inline pkg/handler/iam_handler.go:266]: Remove?

File: pkg/handler/iam_handler_test.go [hunk 7]
```diff
@@ -1102,6 +1102,507 @@ func TestDo(t *testing.T) {
...
+						},
+					},
+				},
+			},
+			wantFoldersPolicy:  &iampb.Policy{},
+			wantProjectsPolicy: &iampb.Policy{},
+			wantErrSubstr:      "failed to check expiry",
+		},
+		{
+			name: "ignore_non-AOD_bindings",
```
Comment (yolocs) [inline pkg/handler/iam_handler_test.go:1382]: How about adding a case for ignoring non-requester bindings?

File: pkg/handler/iam_handler_test.go [hunk 8]
```diff
@@ -1102,6 +1102,507 @@ func TestDo(t *testing.T) {
...
+				},
+			},
+			wantErrSubstr:           "Set IAM policy encountered error: Internal Server Error",
+			wantOrganizationsPolicy: &iampb.Policy{},
+			wantFoldersPolicy: &iampb.Policy{
+				Bindings: []*iampb.Binding{},
+			},
+			wantProjectsPolicy: &iampb.Policy{},
+		},
+	}
```
Comment (yolocs) [inline pkg/handler/iam_handler_test.go:1548]: How about adding a case for mixing multiple bindings together? Expired, not expired, requester bindings, non-requester bindings, etc.

Comment (sqin2019) [reply pkg/handler/iam_handler_test.go:1756]: Added.

File: pkg/handler/iam_handler_test.go (hunk 7 above)
Comment (sqin2019) [reply pkg/handler/iam_handler_test.go:1515]: Added

**Review by yolocs — APPROVED** [review]

**Pull request merged**